Ch-01 What do we mean when we say cabal?

srirajshukla opened the pull request:
This chapter contains the different meanings of the term cabal. 

Right now, I've just added the basic skeleton. I'll be adding more information about the topic.

srirajshukla committed:
Basic skeleton of what do we mean by cabal

Some starter information about the different meanings for the cabal.

srirajshukla commented:
@JonathanLorimer Sorry for being so late on this! I got caught up with some other work. I've started working on it and took some inspiration from your article. 

Is the structure that I'm following right now okay or should I make some changes?

JonathanLorimer commented:
@srirajshukla Hey, thanks for taking this up, and no worries life happens. The structure looks fine for now, I will probably have more to say as you write the sections out.

srirajshukla commented:
Hi @JonathanLorimer, when I am trying to make a commit, I'm getting the error that custom prettier failed even though I've formatted the file with prettier and manually checked with it as well. 

```
bash-4.4$ prettier -c "src/preamble/01_what_do_we_mean_when_we_say_cabal.md"
Checking formatting...
All matched files use Prettier code style!
bash-4.4$ git commit -m "Added details about cabal-install, CABAL, .cabal"
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/sok/.cache/pre-commit/patch1628022078.
Custom Prettier..........................................................Failed
- hook id: custom-prettier
- files were modified by this hook

src/preamble/01_what_do_we_mean_when_we_say_cabal.md

nixpkgs-fmt..........................................(no files to check)Skipped
[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
[INFO] Restored changes from /home/sok/.cache/pre-commit/patch1628022078.
```

Mikolaj commented:
Perhaps do `git add various_files` before `git commit` or do `git commit -a -m "foo"`?

srirajshukla committed:
Added details about cabal-install, CABAL, .cabal

srirajshukla commented:
I got it working, the pre-commit hook was suggesting a change. I wasn't staging it. 😅

JonathanLorimer commented:
@srirajshukla Glad you got it figured out. Sorry for the friction, it can sometimes be a contentious issue whether pre-commit hooks should automatically stage changes or not. Also sorry for the radio silence, I was on vacation. I will have a look at this PR today!

JonathanLorimer commented:
![1628270332_grim](https://user-images.githubusercontent.com/32466011/128548394-81335e79-9834-444d-abf2-80a438068f1e.png)
![1628270340_grim](https://user-images.githubusercontent.com/32466011/128548404-e24729c6-f971-426a-bbd7-8c8c08206513.png)
Just posting some screenshots of running the build locally to demonstrate that things look as though they are styled correctly!

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
 # What do we mean when we say cabal?
+
+Cabal is an umbrella term that can refer to either cabal-the-spec (.cabal
+files), cabal-the-library (code that understands .cabal files), or
+cabal-the-tool (the cabal-install package which provides the cabal executable).
```
Perhaps use quotations rather than hyphens here: "cabal the specification file", "cabal the haskell library", and "cabal the cli tool". I think additional specificity here will pay dividends.

srirajshukla replied:
Yeah, this part should be absolutely clear.

JonathanLorimer replied:
@srirajshukla were you going to change the quoted phrases to be more descriptive?

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+1. [cabal-install](#cabalInstall) is the command line utility to help configure,
+   compile and install Haskell libraries and programs.
+
+2. [The Cabal build system](#cabailBuildSystem), a specification for defining
```
Typo here.
```suggestion
2. [The Cabal build system](#cabalBuildSystem), a specification for defining
```

JonathanLorimer replied:
Is the cabal build system a specification? I think you should also be clear that you are talking about what it commonly referred to as `Cabal` (upper case c) here. I think you are 100% right that it is a library that is used by other haskell packages. It might be worth mentioning that `Cabal` can take a haskell dependency graph (of external and internal modules) and use GHC to build it.

srirajshukla replied:
I'll change Cabal build to CABAL build. I can mention "Cabal can take a haskell dependency graph (of external and internal modules) and use GHC to build it." in the details section?

srirajshukla replied:
@JonathanLorimer  what would be better? adding it here in tldr section or in the more details section?

JonathanLorimer replied:
@srirajshukla I am okay with you mentioning the dependency graph bit in the details section. But I still think you need to be careful about using the word `specification`. It could be confused with the `.cabal` file being a specification for your haskell package. Also I don't think `Cabal` is really a specification, it is really just a library and a build system.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+
+2. [The Cabal build system](#cabailBuildSystem), a specification for defining
+   Haskell packages, together with a library for performing builds. This build
+   system is used by cabal-install and other package managers (stack for eg).
```
I would just write this sentence out.
```suggestion
   system is used by cabal-install and other package managers, stack for example.
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+2. [The Cabal build system](#cabailBuildSystem), a specification for defining
+   Haskell packages, together with a library for performing builds. This build
+   system is used by cabal-install and other package managers (stack for eg).
+   For eg: parsing .cabal files (among others).
```
Can you expand on this?

srirajshukla replied:
I was thinking I'd just add one or two-line summaries here. I'll add some more details here then!

JonathanLorimer replied:
@srirajshukla I am fine with this tl;dr section being small. But I think if you introduce a concept (like parsing a .cabal file) you need to fully flesh it out, otherwise just don't mention it.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+   For eg: parsing .cabal files (among others).
+
+3. [.cabal file](#cabalFile) This is the format that specifies the content of
+   Haskell packages.
```
I think this could be fleshed out more as well. A `.cabal` file specifies a Haskell package with via some top level metadata about the package and build specific information (exposed modules, external dependencies, language extensions, compiler options) about the components that comprise the package.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+3. [.cabal file](#cabalFile) This is the format that specifies the content of
+   Haskell packages.
+
+## The .cabal file<a name="cabalFile"></a>
```
do you need the `<a></a>` tag here? I thought mdbook handled this linking for you. You can test locally using `mdbook serve`

srirajshukla replied:
I am using the <a></a> tags for inline linking. Afaik mdbook handles linking to other files. If there is any other way for linking to topic headers let me know!

JonathanLorimer replied:
mdbook provides these anchor tags for you automatically. You can test what they are by removing your anchor tags, then from the nix shell run `mdbook serve`, navigate to `localhost:3000` and go to the header you want to find the anchor for, click on it and you will see the anchor behind a hashtag in the url.

For example, when I do this for `## The .cabal file` I get this url `http://localhost:3000/preamble/01_what_do_we_mean_when_we_say_cabal.html#the-cabal-file`

Therefore you could just reference `#the-cabal-file` from the tl;dr section!

JonathanLorimer replied:
I think the general rule is that all symbols are stripped and whitespace is replace with hyphens.

srirajshukla replied:
I didn't know it could be done this way.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+
+The .cabal file contains information that drives the compilation and building of
+Haskell packages. This is the format that specifies the content of Haskell
+packages. It is a text-based, key-value format, that is divided into subsections
```
I think you can get rid of this filler sentence
```suggestion
Haskell packages. It is a text-based, key-value format, that is divided into subsections
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+Haskell packages. This is the format that specifies the content of Haskell
+packages. It is a text-based, key-value format, that is divided into subsections
+called stanzas. Each section consists of a number of property descriptions in
+the form of field/value pairs, with a syntax roughly like mail message headers.
```
I think you can remove "in the form of field/value pairs" since you mentioned the `key-value` format above.

I am not sure what you mean by "like mail message headers"

srirajshukla replied:
like mail message header is about email headers, I found the reference in some documentation that I was going through. I think I can remove this whole line. It is not adding anything of much value here.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+packages. It is a text-based, key-value format, that is divided into subsections
+called stanzas. Each section consists of a number of property descriptions in
+the form of field/value pairs, with a syntax roughly like mail message headers.
+It is denoted with a file with the .cabal extension.
```
You can get rid of this sentence since you are referring to `.cabal` files throughout this section.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+the form of field/value pairs, with a syntax roughly like mail message headers.
+It is denoted with a file with the .cabal extension.
+
+This file should contain a number global property descriptions and several
+sections. The package properties describe the package as a whole, such as name,
+license, author, etc. Optionally, a number of configuration flags can be
+declared. These can be used to enable or disable certain features of a package.
+The (optional) library section specifies the library properties and relevant
+build information. Following is an arbitrary number of executable sections which
+describe an executable program and relevant build information.
```
I think this paragraph needs to be tweaked. We need to be very careful about being exceedingly clear here. This is one of the more complicated aspects of cabal.

My suggestions are:
- Start with a sentence explaining what package properties are (perhaps leave out the concept of being "global" they are really just top level fields). Make sure to explain that package properties are metadata about the package that will be built as a result of calling `cabal build`, usually the `.cabal` file lives at the root directory that contains the haskell source code corresponding to the package.
- Explain that the optional component stanzas (library, executable, test-suite, and benchmark) are things called components. Perhaps link to another section of the user guide that covers them in more detail.
- If you are going to mention more specific fields like configuration flags, then make sure to be very clear about what they can do (i.e. enable language extensions, or apply flags to ghc)

srirajshukla replied:
I'll rewrite this paragraph.

srirajshukla replied:
@JonathanLorimer 
I've rewritten this para as: 

The .cabal file contains information that drives the compilation and building of
Haskell packages. Usually the .cabal file lives at the root directory that
contains the haskell source code corresponding to the package. Usually this file
is named as `project-name.cabal`. It is a text-based, key-value format, that is
divided into subsections called stanzas. Each section consists of a number of
property descriptions.

The package properties describe the package as a whole, such as name, license,
author, dependenices, etc. It also contains optional information about optional
components such as
[library properties](../new_to_cabal/06_first_cabal_library.md),
[executables](../new_to_cabal/07_first_cabal_executable.md),
[test-suite](../leveling_up/02_first_cabal_test-suite.md),
[benchmark](src/leveling_up/03_first_cabal_benchmark.md). These components are
also called stanzas.

One of the purposes of Cabal is to make it easier to build a package with
different Haskell implementations. So it provides abstractions of features
present in different Haskell implementations and wherever possible it is best to
take advantage of these to increase portability. For example one of the pieces
of information an author can put in the package’s .cabal file is what language
extensions the code uses. This is far preferable to specifying flags for a
specific compiler as it allows Cabal to pick the right flags for the Haskell
implementation that the user picks. It also allows Cabal to figure out if the
language extension is even supported by the Haskell implementation that the user
picks. Where compiler-specific options are needed however, there is an “escape
hatch” available. The developer can specify implementation-specific options and
more generally there is a configuration mechanism to customise many aspects of
how a package is built depending on the Haskell implementation, the Operating
system, computer architecture and user-specified configuration flags.


Is the last para fine, or should I reduce it?

srirajshukla commented:
@JonathanLorimer Hope you had a great vacation! Yeah, about the pre-commit hook thing, can we add it somewhere (maybe in contributing.md or someplace like that). 

I think I am missing a few references and links, so I'll add them too. And maybe a bit more in the cabal library section.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+build information. Following is an arbitrary number of executable sections which
+describe an executable program and relevant build information.
+
+## The CABAL library <a name="cabalBuildSystem"></a>
```
Same thing about the `<a></a>` tag.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+
+## The CABAL library <a name="cabalBuildSystem"></a>
+
+CABAL stands for Common Architecture for Building Applications and Libraries.
+This is the library that provides functionality that allows the information in
+the .cabal files to be put to use. Without the Cabal library, the Cabal package
+format is just that, a text file. The Cabal library contains the implementations
+that allow for the parsing and operations based on the content of a .cabal file.
+Cabal the library by convention is written with a Capitalization case.
```
Great paragraph. Gets the message across nicely.

JonathanLorimer replied:
```suggestion
CABAL stands for Common Architecture for Building Applications and Libraries.
This is the library that provides functionality that allows the information in
the .cabal files to be put to use. Without the Cabal library, the Cabal package
format is just that, a text file. The Cabal library contains the implementations
that allow for the parsing and operations based on the content of a .cabal file.
Cabal the library by convention is written with Capitalization case.
```
just one small grammatical correction.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+
+## The binary cabal-install (cli tool)<a name="cabalInstall"></a>
+
+The cabal command line tool helps in working with Cabal packages. It (cabal
+lowercase) or more accurately `cabal-install` is the command-line tool that
+provides a user interface for dealing with Haskell packages. cabal-install makes
+use of Cabal the library to do its job.
```
```suggestion
The cabal binary or more accurately `cabal-install` is the command-line tool that
provides a user interface for dealing with Haskell packages. cabal-install makes
use of Cabal the library to do its job.
```

I don't think the first sentence is necessary. Here is a suggested change.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+provides a user interface for dealing with Haskell packages. cabal-install makes
+use of Cabal the library to do its job.
+
+cabal-install is a frontends to CABAL. It makes it possible to build Haskell
```
```suggestion
cabal-install is a frontend to Cabal. It makes it possible to build Haskell
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+use of Cabal the library to do its job.
+
+cabal-install is a frontends to CABAL. It makes it possible to build Haskell
+projects whose sets of dependencies might conflict with each other within the
+confines of a single system. When cabal-install is asked to build a project, by
```
Perhaps you could mention something about version ranges here? https://cabal.readthedocs.io/en/latest/developing-packages.html#modules-imported-from-other-packages

You don't need to go into very much detail, but it might be helpful for the user to understand that they specify a range of acceptable versions and cabal sets out to solve a working package set within those ranges.

srirajshukla replied:
I'll add a bit about it here.

srirajshukla replied:
@JonathanLorimer  I've added this much: 
"Packages provide a constraint on the version of
various libraries (version range) that will work with their package."

Is this enough?

JonathanLorimer replied:
Yeah that is fine, maybe add something like "Packages provide a constraint on the version of
various libraries (using cabal's version bounds syntax) that will work with their package."

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+confines of a single system. When cabal-install is asked to build a project, by
+default it looks at the dependencies specified in its .cabal file and uses a
+dependency solver to figure out a set of packages and package versions that
+satisfy it. This set of package is drawn from all package and all versions from
```
```suggestion
satisfy it. This set of packages is drawn from all package and all versions from
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+Hackage as a whole. The chosen versions of the dependenices will be installed
+and indexed in a database in the cabal directory.
+
+Version conflicts between dependencies are avoided by indexing the installed
+packages according to their version and other relevent configuration options.
+This allows different projects to retrieve the dependecy versions they need.
```
```suggestion
Conflicts between dependencies are avoided by indexing the installed
packages according to their version and other relevant configuration options.
This allows different projects to retrieve the dependency versions they need.
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+packages according to their version and other relevent configuration options.
+This allows different projects to retrieve the dependecy versions they need.
+
+    stack is also a command-line tool that depends on the Cabal Library,
+    and hence also consumes the information specified in Cabal Package format
+    found in the .cabal files.
```
Is the indentation here intended?

srirajshukla replied:
Not really.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,73 @@
+
+# References
+
+1. [CABAL hackage page](https://hackage.haskell.org/package/Cabal)
```
```suggestion
1. [Cabal hackage page](https://hackage.haskell.org/package/Cabal)
```

JonathanLorimer reviewed (changes requested):
This is a great start. Thanks again for your contribution. I know I left a lot of comments but I think this is a really great first attempt. Please ping me if you have any questions!

srirajshukla committed:
Changes according to suggestions

1. Rewrote about `.cabal` section
2. Changed tl;dr for cabal library and added a bit more about it.
3. Changed inline href from anchor tags to direct linking.
4. Fixed typos.

srirajshukla commented:
@JonathanLorimer

JonathanLorimer commented:
@srirajshukla Thanks for getting these changes in so quickly! I will give this a review tonight.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+1. [cabal-install](#the-cabal-file) is the command line utility to help
+   configure, compile and install Haskell libraries and programs.
+
+2. [The Cabal build system](#the-cabal-library), a library that is used for
```
Should there be a comma at the end of the link here? There isn't on `cabal-install` and `.cabal file`

srirajshukla replied:
"The cabal build system is a library that is used" would be better, yeah.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+
+## tl;dr
+
+1. [cabal-install](#the-cabal-file) is the command line utility to help
```
Why does `cabal-install` link to `#the-cabal-file`?

srirajshukla replied:
Oops, linked to the wrong para. I put it in the order the paragraphs were, and tl;dr list is in different order. That's why. I'll rearrange them, and link to proper sections.

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+   used by cabal-install and other package manager, stack for example. For eg:
+   parsing .cabal files (among others).
+
+3. [.cabal file](#the-binary-cabal-install-cli-tool) This is the format that
```
Seems like this is the wrong link as well.

JonathanLorimer replied:
Also I would say its "the file" not "the format" that specifies the content of a Haskell package. Perhaps you mean something like "It is a file with a custom format designed for specifying the content of Haskell packages"?

srirajshukla replied:
Yes, it should be "the file".

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+   configure, compile and install Haskell libraries and programs.
+
+2. [The Cabal build system](#the-cabal-library), a library that is used for
+   reation of packages and the building of their contents. This build system is
```
Missing a c on creation, and switch from passive to active voice.
```suggestion
   creating packages and building their contents. This build system is
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+
+2. [The Cabal build system](#the-cabal-library), a library that is used for
+   reation of packages and the building of their contents. This build system is
+   used by cabal-install and other package manager, stack for example. For eg:
```
```suggestion
   used by cabal-install and other package managers, stack for example. For eg:
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+2. [The Cabal build system](#the-cabal-library), a library that is used for
+   reation of packages and the building of their contents. This build system is
+   used by cabal-install and other package manager, stack for example. For eg:
+   parsing .cabal files (among others).
```
Again, can we get rid of `For eg:` its a little too informal for a guide. If you are going to explain what some of libraries capabilities are you should list more than just parsing (I usually cite 3 examples as a rule of thumb).

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+   specifies the content of Haskell packages. A .cabal file specifies a Haskell
+   package with via some top level metadata about the package and build specific
+   information (exposed modules, external dependencies, language extensions,
+   compiler options) about the components that comprise the package.
```
A couple typos and some word repetition, perhaps something like this? 
```suggestion
   specifies the content of Haskell packages. The types of information that can be provided via the
   bespoke file format are: top level metadata about the package, and build specific
   information (exposed modules, external dependencies, language extensions,
   compiler options) about the components that comprise the package.
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+The .cabal file contains information that drives the compilation and building of
+Haskell packages. Usually the .cabal file lives at the root directory that
+contains the haskell source code corresponding to the package. Usually this file
+is named as `project-name.cabal`. It is a text-based, key-value format, that is
```
Perhaps add angle brackets to emphasize that `project-name` is just a place holder.
```suggestion
is named as `<project-name>.cabal`. It is a text-based, key-value format, that is
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+[library properties](../new_to_cabal/06_first_cabal_library.md),
+[executables](../new_to_cabal/07_first_cabal_executable.md),
+[test-suite](../leveling_up/02_first_cabal_test-suite.md),
+[benchmark](src/leveling_up/03_first_cabal_benchmark.md). These components are
+also called stanzas.
```
Maybe just remove the last sentence, it feels a bit redundant.
```suggestion
[benchmark](src/leveling_up/03_first_cabal_benchmark.md).
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+present in different Haskell implementations and wherever possible it is best to
+take advantage of these to increase portability. For example one of the pieces
+of information an author can put in the package’s .cabal file is what language
+extensions the code uses. This is far preferable to specifying flags for a
```
```suggestion
extensions the code uses. This is preferable to specifying flags for a
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+specific compiler as it allows Cabal to pick the right flags for the Haskell
+implementation that the user picks. It also allows Cabal to figure out if the
+language extension is even supported by the Haskell implementation that the user
+picks. Where compiler-specific options are needed however, there is an “escape
```
```suggestion
picks. Where compiler-specific options are needed, there is an “escape
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+hatch” available. The developer can specify implementation-specific options and
+more generally there is a configuration mechanism to customise many aspects of
+how a package is built depending on the Haskell implementation, the Operating
+system, computer architecture and user-specified configuration flags.
```
Everything should be lower case
```suggestion
how a package is built depending on the Haskell implementation, the operating
system, computer architecture and user-specified configuration flags.
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+more generally there is a configuration mechanism to customise many aspects of
+how a package is built depending on the Haskell implementation, the Operating
+system, computer architecture and user-specified configuration flags.
+
```
This section is _really_ fantastic. Well written!

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+
+## The CABAL library
+
+CABAL stands for Common Architecture for Building Applications and Libraries.
```
```suggestion
Cabal stands for Common Architecture for Building Applications and Libraries.
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+CABAL stands for Common Architecture for Building Applications and Libraries.
+This is the library that provides functionality that allows the information in
+the .cabal files to be put to use. Without the Cabal library, the Cabal package
+format is just that, a text file. The Cabal library contains the implementations
```
I think this sentence is unnecessary.
```suggestion
the .cabal files to be put to use. The Cabal library contains the implementations
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+the .cabal files to be put to use. Without the Cabal library, the Cabal package
+format is just that, a text file. The Cabal library contains the implementations
+that allow for the parsing and operations based on the content of a .cabal file.
+Cabal the library by convention is written with Capitalization case. Cabal can
```
```suggestion
Cabal the library, by convention, is written with Capitalization case. Cabal can
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+cabal-install is a frontend to Cabal. It makes it possible to build Haskell
+projects whose sets of dependencies might conflict with each other within the
+confines of a single system. Packages provide a constraint on the version of
+various libraries (version range) that will work with their package. When
```
```suggestion
confines of a single system. A package's .cabal file provides a constraints on the version of
various libraries (version bounds) that they expect. When
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+cabal-install is asked to build a project, by default it looks at the
+dependencies specified in its .cabal file and uses a dependency solver to figure
+out a set of packages and package versions that satisfy it. This set of packages
+is drawn from all package and all versions from Hackage as a whole. The chosen
```
```suggestion
is drawn from Hackage. The chosen
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,95 @@
+out a set of packages and package versions that satisfy it. This set of packages
+is drawn from all package and all versions from Hackage as a whole. The chosen
+versions of the dependenices will be installed and indexed in a database in the
+cabal directory.
```
do you mean `~/.cabal` directory? I am pretty sure that's the correct one. It would be good to confirm this in the cabal docs either way.

srirajshukla committed:
Fixed wrong links, reworded some parts.

srirajshukla commented:
@JonathanLorimer I was sick for a few days so I didn't do it earlier. I've made the changes.

srirajshukla commented:
Hi @JonathanLorimer , please review it when you've got some time

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,92 @@
+
+## The .cabal file
+
+The .cabal file contains information that drives the compilation and building of
+Haskell packages. Usually the .cabal file lives at the root directory that
+contains the haskell source code corresponding to the package. Usually this file
+is named as `<project-name>.cabal`. It is a text-based, key-value format, that
+is divided into subsections called stanzas.
```
```suggestion
The .cabal file contains information that drives the compilation and building of
Haskell packages. The .cabal file lives at the root directory that
contains the haskell source code corresponding to the package. By convention this file
is named `<project-name>.cabal`. It is a text-based, key-value format, that
is divided into subsections called stanzas.
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,92 @@
+present in different Haskell implementations and wherever possible it is best to
+take advantage of these to increase portability. For example one of the pieces
+of information an author can put in the package’s .cabal file is what language
+extensions the code uses. This is preferable to specifying flags for a specific
+compiler as it allows Cabal to pick the right flags for the Haskell
+implementation that the user picks. It also allows Cabal to figure out if the
+language extension is even supported by the Haskell implementation that the user
+picks. Where compiler-specific options are needed, there is an “escape hatch”
+available. The developer can specify implementation-specific options and more
+generally there is a configuration mechanism to customise many aspects of how a
+package is built depending on the Haskell implementation, the operating system,
+computer architecture and user-specified configuration flags.
```
Perhaps it is worth mentioning that these days, we have pretty much all converged on GHC as the standard haskell implementation.

srirajshukla replied:
@JonathanLorimer Where will it be suitable to mention? 

**At the beginning like this:** 
One of the purposes of Cabal is to make it easier to build a package with different Haskell implementations. There are various [Haskell implementations](https://wiki.haskell.org/Implementations), but we have now pretty much all converged on [GHC](https://www.haskell.org/ghc/) as the standard Haskell implementation. So it provides ...

**Or end the end:**

... user-specified configuration flags. There are various [Haskell implementations](https://wiki.haskell.org/Implementations), but we have now pretty much all converged on [GHC](https://www.haskell.org/ghc/) as the standard Haskell implementation.

I'm leaning more towards the latter option.

JonathanLorimer replied:
Yeah, I agree that the end makes the most sense

srirajshukla replied:
ok cool!

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,92 @@
+
+## The CABAL library
+
+Cabal stands for Common Architecture for Building Applications and Libraries.
+This is the library that provides functionality that allows the information in
+the .cabal files to be put to use. The Cabal library contains the
+implementations that allow for the parsing and operations based on the content
+of a .cabal file. Cabal the library, by convention, is written with
+Capitalization case. Cabal can take a haskell dependency graph (of external and
+internal modules) and use GHC to build it.
```
```suggestion
Cabal stands for Common Architecture for Building Applications and Libraries.
This is the library that provides functionality that allows the information in
the .cabal files to be put to use. The Cabal library contains the
code for parsing .cabal files and operations for building
haskell packages. Cabal the library, by convention, is written with
Capitalization case. Cabal can take a haskell dependency graph (of external and
internal modules) and use GHC to build it.
```

JonathanLorimer commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,92 @@
+
+cabal-install is a frontend to Cabal. It makes it possible to build Haskell
+projects whose sets of dependencies might conflict with each other within the
+confines of a single system. A package's .cabal file provides a constraints on
```
```suggestion
confines of a single system. A package's .cabal file provides a constraint on
```

JonathanLorimer reviewed (commented):
If you include a couple of the suggestions I left, I am happy to approve this. Thanks for the contribution!

srirajshukla committed:
Fixed typos, grammar mistakes, added more info

1. Fixed some typos pointed out in code review.
2. Added more information on Haskell implementations.

srirajshukla committed:
Fixed typos, grammar mistakes, added more info

1. Fixed some typos pointed out in code review.
2. Added more information on Haskell implementations.

srirajshukla commented:
@JonathanLorimer  I hope it is fine now. 😅

JonathanLorimer commented:
@Mikolaj can you give this a review?

Mikolaj commented:
> @Mikolaj can you give this a review?

Not right now, but yes, with pleasure.

jhrcek commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,97 @@
+The package properties describe the package as a whole, such as name, license,
+author, dependenices, etc. It also contains optional information about optional
+components such as
+[library properties](../new_to_cabal/06_first_cabal_library.md),
```
Nitpick: 
"about optional components such as library properties" - sounds weird. Is "library property" an optional component?
I'd change this to `[libraries]`. Also in the "optional information about optional components" I'd drop the first "optional".
Since the component is optional, it means it can be left out, so no need to mention the optionality twice.

srirajshukla replied:
Yep, agree. I'll remove the first optional.

jhrcek reviewed (dismissed):
Nice work :+1:

srirajshukla committed:
Merge branch 'chapter-01' of https://github.com/srirajshukla/cabal-userguide into chapter-01

Mikolaj commented:
I apologise for the delay in reviewing --- we've just had regressions discovered in the published 3.6 and also CI got completely broken by changes in GitHub Actions. We are still saving the world, but I will review later, unless somebody else does and unblocks you.

Mikolaj commented:
OK, I'm back, will review RSN. :)

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
 # What do we mean when we say cabal?
+
+Cabal is an umbrella term that can refer to either "cabal the spec" (.cabal
```
"the package spec"? I wasn't sure if you mean "specification of the cabal software system" or "cabal format spec" or "cabal protocol spec".

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+## tl;dr
+
+1. [.cabal file](#the-cabal-file) This is the file that specifies the content of
+   Haskell packages. The types of information that can be provided via the
```
I'd cross out "that can be". There is not so much optionality here, no defaults, a lot of the info is mandatory.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+   cabal-install and other package managers, stack for example.
+
+3. [cabal-install](#the-binary-cabal-install-cli-tool) is the command line
+   utility to help configure, compile and install Haskell libraries and
```
I'd change "to help configure" to "for configuring", or something similar, because cabal-install is not a helper that makes it easier. There are no ways to do that except via cabal-inistall.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+## The .cabal file
+
+The .cabal file contains information that drives the compilation and building of
+Haskell packages. The .cabal file lives at the root directory that contains the
```
I'm not a native speaker, but I suspect "in root directory".

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+
+The .cabal file contains information that drives the compilation and building of
+Haskell packages. The .cabal file lives at the root directory that contains the
+haskell source code corresponding to the package. By convention this file is
```
I'd rephrase to "in root directory of the source tree of the package". Root directory does not contain code. It contains other directories that contain code and other directories, etc.

Mikolaj replied:
I'm not a native speaker, so I'm not sure about 'convention'. This is mandatory, so surely not a social convention, but built into the format. Not sure how to better express that.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+The .cabal file contains information that drives the compilation and building of
+Haskell packages. The .cabal file lives at the root directory that contains the
+haskell source code corresponding to the package. By convention this file is
+named `<project-name>.cabal`. It is a text-based, key-value format, that is
```
"It is in a ,.. format". We are still talking about the file, not its format. So file is in format. Format is text-based.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+
+The package properties describe the package as a whole, such as name, license,
+author, dependenices, etc. It also contains information about optional
+components such as [library](../new_to_cabal/06_first_cabal_library.md),
```
Either 'libraries' or reword, because a single library is mandatory, everything else is optional.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+author, dependenices, etc. It also contains information about optional
+components such as [library](../new_to_cabal/06_first_cabal_library.md),
+[executables](../new_to_cabal/07_first_cabal_executable.md),
+[test-suite](../leveling_up/02_first_cabal_test-suite.md),
+[benchmark](src/leveling_up/03_first_cabal_benchmark.md). These components are
```
test-suites, benchmarks; they can be many, as well.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+also called stanzas.
+
+One of the purposes of Cabal is to make it easier to build a package with
+different Haskell implementations. So it provides abstractions of features
```
Given that currently GHC is virtually the only one, I'd clarify "(e.g., different versions of the GHC compiler or one of the past or future competitors)".  Because there is no present competitor, sadly. :)

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+
+One of the purposes of Cabal is to make it easier to build a package with
+different Haskell implementations. So it provides abstractions of features
+present in different Haskell implementations and wherever possible it is best to
+take advantage of these to increase portability. For example one of the pieces
```
I'd be a little less proscriptive and condense to "present in different Haskell implementations that can be used to increase portability.".

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+package is built depending on the Haskell implementation, the operating system,
+computer architecture and user-specified configuration flags.
+
+There are various
+[Haskell implementations](https://wiki.haskell.org/Implementations), but we have
+now pretty much all converged on [GHC](https://www.haskell.org/ghc/) as the
+standard Haskell implementation.
+
```
If we mention GHC above, as suggested, I'd remove this. Perhaps re-use the links, GHC for GHC, Implementations for "past and future competitors".

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+## The CABAL library
+
+Cabal stands for Common Architecture for Building Applications and Libraries.
+This is the library that provides functionality that allows the information in
+the .cabal files to be put to use. The Cabal library contains the code for
```
I bet you could condense this sentence to half the length. :)

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+This is the library that provides functionality that allows the information in
+the .cabal files to be put to use. The Cabal library contains the code for
+parsing .cabal files and operations for building haskell packages. Cabal the
+library, by convention, is written with Capitalization case. Cabal can take a
```
I'm not a native speaker, so I'd need to google to understand what "Capitalization case" means.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+This is the library that provides functionality that allows the information in
+the .cabal files to be put to use. The Cabal library contains the code for
+parsing .cabal files and operations for building haskell packages. Cabal the
+library, by convention, is written with Capitalization case. Cabal can take a
+haskell dependency graph (of external and internal modules) and use GHC to build
+it.
```
I'd say "Cabal generates a dependency graph (of external and internal packages and modules) and repeatedly calls GHC to build it."

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+haskell dependency graph (of external and internal modules) and use GHC to build
+it.
+
+## The binary cabal-install (cli tool)
```
This is hard, because the exe is not called `cabal-install` and we don't want to mislead the users. How about "The commandline tool (cabal-instal)"? I'm not happy with that either.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+
+## The binary cabal-install (cli tool)
+
+The cabal binary or more accurately `cabal-install` is the command-line tool
```
I'd repeat the section title: "The commandline tool project called `cabal-install`, with the executable confusingly named `cabal`, provides user interface..."

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+## The binary cabal-install (cli tool)
+
+The cabal binary or more accurately `cabal-install` is the command-line tool
+that provides a user interface for dealing with Haskell packages. cabal-install
+makes use of Cabal the library to do its job.
+
+cabal-install is a frontend to Cabal. It makes it possible to build Haskell
```
These two sentences are almost repeated.

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+that provides a user interface for dealing with Haskell packages. cabal-install
+makes use of Cabal the library to do its job.
+
+cabal-install is a frontend to Cabal. It makes it possible to build Haskell
+projects whose sets of dependencies might conflict with each other within the
+confines of a single system. A package's .cabal file provides a constraint on
```
This is important point, but hard to get across. Perhaps "<something about cabal-install being powerful thanks to using Cabal library> It's perfectly comfortable building two Haskell projects whose sets of dependencies are conflicting and storing their artefacts to be concurrently employed by the user."

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+the version of various libraries (version bounds) that they expect. When
+cabal-install is asked to build a project, by default it looks at the
+dependencies specified in its .cabal file and uses a dependency solver to figure
+out a set of packages and package versions that satisfy it. This set of packages
```
I'd say "packages with their versions" to be able to later on omit "and package versions".

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+cabal-install is asked to build a project, by default it looks at the
+dependencies specified in its .cabal file and uses a dependency solver to figure
+out a set of packages and package versions that satisfy it. This set of packages
+is drawn from Hackage. The chosen versions of the dependenices will be installed
```
"Hackage (the central Haskell package repository) or another repository specified by the user"; BTW, typo: dependenices

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+
+Conflicts between dependencies are avoided by indexing the installed packages
+according to their version and other relevant configuration options. This allows
+different projects to retrieve the dependency versions they need.
```
I'd add "dependency versions and build flavours".

Mikolaj commented on src/preamble/01_what_do_we_mean_when_we_say_cabal.md:
```diff
@@ -1 +1,96 @@
+according to their version and other relevant configuration options. This allows
+different projects to retrieve the dependency versions they need.
+
+stack is also a command-line tool that depends on the Cabal Library, and hence
```
Let's be high-minded and also compliment ourselves saying something like "stack is another mature command-line tool, with somewhat different goals and trade-offs, that depends...".

Mikolaj reviewed (approved):
I like the choice of material, technical accuracy and style. Good job!
